### PR TITLE
Controller

### PR DESCRIPTION
**Pull request by rtsmoor**
controller code with the BFS (Breadth-First-Search) and hungarian algorithm. Merging so that this code can be used on the server (controller->server branch)

### CODE REVIEW

**Review by intigames123 — APPROVED** [review]

File: controller.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+var target;
+
+for (let firstTargetSetIndex = 0; firstTargetSetIndex < firstTargetSet.length; firstTargetSetIndex++) {
+    target = firstTargetSet[firstTargetSetIndex];
+    targetSets[0].push(target);
+
+    for (let targetSetIndex = 0; targetSetIndex < numberOfTargetSets - 1; targetSetIndex++) {
+    target += targetSetPattern[targetSetIndex % targetSetPattern.length];
+    targetSets[targetSetIndex + 1].push(target);
+    }
```
Comment (rtsmoor) [inline controller.js:33]: Indentation

File: controller.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+            }
+            else if (currentRow == 0 && currentColumn != columns - 1) { // TOP CENTER
+                adjacencyMatrix[index][rightCel] = 1;
+                adjacencyMatrix[index][celDown] = 1;
+                adjacencyMatrix[index][leftCel] = 1;
+            }
+            else if (currentRow == 0 && currentColumn == columns - 1) { // TOP RIGHT
+                adjacencyMatrix[index][celDown] = 1;
+                adjacencyMatrix[index][leftCel] = 1;
+            }
```
Comment (rtsmoor) [inline controller.js:65]:
```suggestion
            if(currentRow == 0){
                    if (currentRow == 0 && currentColumn == 0) { // TOP LEFT
                        adjacencyMatrix[index][rightCel] = 1;
                        adjacencyMatrix[index][celDown] = 1;
                    }
                    else if (currentRow == 0 && currentColumn != columns - 1) { // TOP CENTER
                        adjacencyMatrix[index][rightCel] = 1;
                        adjacencyMatrix[index][celDown] = 1;
                        adjacencyMatrix[index][leftCel] = 1;
                    }
                    else if (currentRow == 0 && currentColumn == columns - 1) { // TOP RIGHT
                        adjacencyMatrix[index][celDown] = 1;
                        adjacencyMatrix[index][leftCel] = 1;
                    }
            }
```

File: controller.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                adjacencyMatrix[index][celUp] = 1;
+                adjacencyMatrix[index][rightCel] = 1;
+                adjacencyMatrix[index][celDown] = 1;
+                adjacencyMatrix[index][leftCel] = 1;
+            }
+            else if ((currentRow > 0 && currentRow < rows - 1) && currentColumn == columns - 1) { // CENTER RIGHT
+                adjacencyMatrix[index][celUp] = 1;
+                adjacencyMatrix[index][leftCel] = 1;
+                adjacencyMatrix[index][celDown] = 1;
+            }
```
Comment (rtsmoor) [inline controller.js:81]:
```suggestion
            else if(currentRow > 0){
                        else if ((currentRow < rows - 1) && currentColumn == 0) { // CENTER LEFT
                            adjacencyMatrix[index][celUp] = 1;
                            adjacencyMatrix[index][rightCel] = 1;
                            adjacencyMatrix[index][celDown] = 1;
                        }
                        else if ((currentRow < rows - 1) && (currentColumn > 0 && currentColumn < columns - 1)) { // CENTER CENTER
                            adjacencyMatrix[index][celUp] = 1;
                            adjacencyMatrix[index][rightCel] = 1;
                            adjacencyMatrix[index][celDown] = 1;
                            adjacencyMatrix[index][leftCel] = 1;
                        }
                        else if ((currentRow < rows - 1) && currentColumn == columns - 1) { // CENTER RIGHT
                            adjacencyMatrix[index][celUp] = 1;
                            adjacencyMatrix[index][leftCel] = 1;
                            adjacencyMatrix[index][celDown] = 1;
                        }
            }
```

File: controller.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            else if (currentRow == rows - 1 && (currentColumn > 0 && currentColumn < columns - 1)) { // BOTTOM CENTER
+                adjacencyMatrix[index][celUp] = 1;
+                adjacencyMatrix[index][rightCel] = 1;
+                adjacencyMatrix[index][leftCel] = 1;
+            }
+            else if (currentRow == rows - 1 && currentColumn == columns - 1) { // BOTTOM RIGHT
+                adjacencyMatrix[index][celUp] = 1;
+                adjacencyMatrix[index][leftCel] = 1;
+            }
+        }
```
Comment (rtsmoor) [inline controller.js:95]:
```suggestion
            else{
                else if (currentRow == rows - 1 && currentColumn == 0) { // BOTTOM LEFT
                    adjacencyMatrix[index][celUp] = 1;
                    adjacencyMatrix[index][rightCel] = 1;
                }
                else if (currentRow == rows - 1 && (currentColumn > 0 && currentColumn < columns - 1)) { // BOTTOM CENTER
                    adjacencyMatrix[index][celUp] = 1;
                    adjacencyMatrix[index][rightCel] = 1;
                    adjacencyMatrix[index][leftCel] = 1;
                }
                else if (currentRow == rows - 1 && currentColumn == columns - 1) { // BOTTOM RIGHT
                    adjacencyMatrix[index][celUp] = 1;
                    adjacencyMatrix[index][leftCel] = 1;
                }
            }
        }
```

**Pull request merged**